src/modules/borrow-history/borrowhistory.service.ts

```diff
@@ -0,0 +1,147 @@
+import { Book } from '../book/domain/book';
+
+@Injectable()
+export class BorrowHistoryService {
+  constructor(
+    @InjectRepository(BorrowHistoryEntity)
+    private readonly borrowHistoryRepository: Repository<BorrowHistoryEntity>,
+
+    @InjectRepository(BookEntity)
+    private readonly bookRepository: Repository<BookEntity>,
+
+    @InjectRepository(UserEntity)
+    private readonly userRepository: Repository<UserEntity>,
+  ) {}
+
+  async findAll(): Promise<BorrowHistory[]> {
+    return BorrowHistory.fromEntities(
+      await this.borrowHistoryRepository.find(),
+    );
+  }
+
+  async findById(id: number): Promise<BorrowHistory> {
+    return BorrowHistory.fromEntity(await this.findOneOrThrow(id));
+  }
+
+  async create(borrowBook: BorrowBook): Promise<BorrowHistory> {
+    const book = await this.findBookOrThrow(borrowBook.bookId);
+    const user = await this.findUserOrThrow(borrowBook.userId);
+
+    const entity = await this.borrowHistoryRepository.save(
+      this.borrowHistoryRepository.create({
+        user,
+        book,
+        userId: user.id,
+        bookId: book.id,
+        borrowDate: new Date(),
+      }),
+    );
+
+    await this.bookRepository.save({ ...book, available: false });
+
+    return BorrowHistory.fromEntity(entity);
+  }
+
+  async remove(id: number): Promise<BorrowHistory> {
+    const entity = await this.findOneOrThrow(id);
+
+    if (entity.status === BorrowStatusEnum.RETURNED) {
+      throw new Error(`Book with borrow history ID ${id} is already returned`);
+    }
+
+    const updatedEntity = await this.borrowHistoryRepository.save({
+      ...entity,
+      returnDate: new Date(),
+      status: BorrowStatusEnum.RETURNED,
+    });
+
+    const book = entity.book;
+    if (!book) {
+      throw new NotFoundException(`Book not found in borrow history`);
+    }
+
+    await this.bookRepository.save({ ...book, available: true });
+
+    return BorrowHistory.fromEntity(updatedEntity);
+  }
+
+  async getBorrowHistoryByCodeEan13(
+    code: string,
+  ): Promise<{ available: boolean; borrowHistories: BorrowHistory[] }> {
+    const bookEntity = await this.bookRepository.findOneBy({ code });
+
+    if (!bookEntity) {
+      throw new NotFoundException(`Book with code ${code} not found`);
+    }
+
+    const borrowHistoryEntities = await this.borrowHistoryRepository.find({
+      where: { bookId: bookEntity.id },
+      relations: ['user', 'book'],
+      order: { borrowDate: 'DESC' },
+    });
+
+    return {
+      available: bookEntity.available,
+      borrowHistories: BorrowHistory.fromEntities(borrowHistoryEntities),
+    };
+  }
+  private async findOneOrThrow(id: number): Promise<BorrowHistoryEntity> {
+    const entity = await this.borrowHistoryRepository.findOne({
+      where: { id },
+      relations: ['book', 'user'],
+    });
+    if (!entity) {
+      throw new NotFoundException(`Borrow history with ID ${id} not found`);
+    }
```

Comment:
Trừ khi dùng query builder, còn lại luôn phải specific cột theo tên reference, không dùng string
Chú ý dòng trống


```suggestion
    const entity = await this.borrowHistoryRepository.findOne({
      where: { id },
      relations: {
         book: true,
         user: true
      },
    });
    
    if (!entity) {
      throw new NotFoundException(`Borrow history with ID ${id} not found`);
    }
```